statix.solver/src/main/java/mb/statix/spec/ARule.java

```diff
@@ -240,18 +240,19 @@ public static Optional<Integer> compare(Rule r1, Rule r2) {
             return Pattern.leftRightOrdering.compare(p1, p2);
         }
 
-        public Comparator<Rule> asComparator() {
+        public static Comparator<Rule> asComparator() {
             return (r1, r2) -> LeftRightOrder.compare(r1, r2).orElse(0);
         }
 
     }
 
     public static class LeftRightOrderWithConsistentEquality {
         public Comparator<Rule> asComparator() {
-            return (r1, r2) -> {
-                final Optional<Integer> patternComparison = LeftRightOrder.compare(r1, r2);
-                return patternComparison.filter(i -> i != 0).orElse(Integer.compare(r1.hashCode(), r2.hashCode()));
-            };
+            return LeftRightOrder.asComparator()
+                .thenComparing((r1, r2) -> Integer.compare(r1.hashCode(), r2.hashCode()))
+                // hash collision handling...
```

Comment:
Also this version does not handle hash collisions correctly. I'm afraid the only sound way is comparing all members of `Rule` one-by-one.